pinot-spi/src/main/java/org/apache/pinot/spi/utils/ByteArray.java

```diff
@@ -91,19 +109,55 @@ public int compareTo(@Nonnull ByteArray that) {
    *   <li> +ve integer if first value is larger than the second. </li>
    * </ul>
    *
-   * @param bytes1 First byte[] to compare.
-   * @param bytes2 Second byte[] to compare.
+   * @param left First byte[] to compare.
+   * @param right Second byte[] to compare.
+   * @return Result of comparison as stated above.
+   */
+  public static int compare(byte[] left, byte[] right) {
+    return compare(left, 0, left.length, right, 0, right.length);
+  }
+
+  /**
+   * Compares two byte[] values. The comparison performed is on unsigned value for each byte.
+   * Returns:
+   * <ul>
+   *   <li> 0 if both values are identical. </li>
+   *   <li> -ve integer if first value is smaller than the second. </li>
+   *   <li> +ve integer if first value is larger than the second. </li>
+   * </ul>
+   *
+   * @param left First byte[] to compare.
+   * @param leftFromIndex inclusive index of first byte to compare in left
+   * @param leftToIndex exclusive index of last byte to compare in left
+   * @param right Second byte[] to compare.
+   * @param rightFromIndex inclusive index of first byte to compare in right
+   * @param rightToIndex exclusive index of last byte to compare in right
    * @return Result of comparison as stated above.
    */
-  public static int compare(byte[] bytes1, byte[] bytes2) {
-    int len1 = bytes1.length;
-    int len2 = bytes2.length;
+  public static int compare(byte[] left, int leftFromIndex, int leftToIndex,
+      byte[] right, int rightFromIndex, int rightToIndex) {
+    if (COMPARE_UNSIGNED != null) {
+      try {
+        return (int) COMPARE_UNSIGNED.invokeExact(left, leftFromIndex, leftToIndex,
+            right, rightFromIndex, rightToIndex);
+      } catch (ArrayIndexOutOfBoundsException outOfBounds) {
+        throw outOfBounds;
+      } catch (Throwable ignore) {
```

Comment:
when would this happen, should we propagate it

Reply:
it can't happen, it's just a consequence of the API. Propagating `ArrayIndexOutOfBoundsException` is important though.

Reply:
can we add tests for the IOE and fallback for JDK9+?

Reply:
Good call - turns out we don't have a test for this class at all 😬

Reply:
tests here https://github.com/apache/pinot/pull/7487/commits/efa0e25504391f19ce6e76c4357cfdf6be666664